share/getters/shrex.go

```diff
@@ -39,22 +38,23 @@ type metrics struct {
 	ndAttempts  syncint64.Histogram
 }
 
-func (m *metrics) recordEDSAttempt(attemptCount int, success bool) {
+func (m *metrics) recordEDSAttempt(ctx context.Context, attemptCount int, success bool) {
 	if m == nil {
 		return
 	}
-	ctx, cancel := context.WithTimeout(context.Background(), metricObservationTimeout)
-	defer cancel()
+	if ctx.Err() != nil {
+		ctx = context.Background()
+	}
 	m.edsAttempts.Record(ctx, int64(attemptCount), attribute.Bool("success", success))
 }
 
-func (m *metrics) recordNDAttempt(attemptCount int, success bool) {
+func (m *metrics) recordNDAttempt(ctx context.Context, attemptCount int, success bool) {
 	if m == nil {
 		return
 	}
-
-	ctx, cancel := context.WithTimeout(context.Background(), metricObservationTimeout)
-	defer cancel()
+	if ctx.Err() != nil {
+		ctx = context.Background()
+	}
 	m.ndAttempts.Record(ctx, int64(attemptCount), attribute.Bool("success", success))
```

Comment:
@walldiss, an in code explanation
```suggestion
	if ctx.Err() != nil { // cancelled 
		ctx = context.Background() // we ignore that it is cancelled 
	}
	m.ndAttempts.Record(ctx, int64(attemptCount), attribute.Bool("success", success)) // and record success anyway, when it's not
```

Reply:
In your example `success` is a var and it is passed from caller, allowing him to control submitted value as he want.

Reply:
Resolved then